### PR TITLE
Instructions for rookies + install-linux target in Makefile

### PR DESCRIPTION
**Pull request by chenyow**
Detailed steps on how to build and install the firmware for Debian and Ubuntu systems.

### CODE REVIEW

**Comment by Ichimonji10** [comment]
:+1:

**Comment by olerem** [comment]
Hi,
 I have some doubts about make install-linux part:
- the back up will be overwritten on second install,
- access rights for new files are wrong.

User should copy original firmware to save place. I know no installer which cares about copying target file to backup.
Probably will be better if you remove make patch and just describes needed copy + chown steps in documentation.

**Comment by chenyow** [comment]
Both of your comments are incorrect:
* ~~the back up will be overwritten on second install.~~
*Incorrect:* cp uses -n (no clobber) option, the backup will not be overwritten if it already exists.
* ~~access rights for new files are wrong.~~
*Incorrect:* both the backup and the new firmware are copied to /lib/firmware as root:root with the correct permissions, because install-linux is run under sudo. If the user runs the command without sudo, no files are touched / changed and a permission error is displayed, which is the correct behavior.

In short, this is perfectly good change request, and exactly what a rookie needs to safely install new firmware. Instead you're asking the user to RTFM chown and friends, and understand the intricacies of /lib/firmware and handling root-owned files.

**Comment by olerem** [comment]
You right, my mistake.

**Comment by olerem** [comment]
Please add comments to your patches and add Signed-off-by:.

**Comment by olerem** [comment]
After this patch https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=e904cf6fe23022cde4e0ea9d41601411a315a3dc (kernel v4.4-rc1) we use file name as FW version.
Most of this patch need to be updated. @chenyow would you like to update it?

**Comment by rfelten** [comment]
+1

**Comment by DukhiAatma** [comment]
hi sir, after following the steps the following errors appear

root@debian:~/Desktop# dmesg | tail
[ 3841.784928] usb 1-1: new full-speed USB device number 6 using ohci-pci
[ 3842.251620] usb 1-1: config 1 interface 0 altsetting 0 endpoint 0x1 has invalid maxpacket 512, setting to 64
[ 3842.251623] usb 1-1: config 1 interface 0 altsetting 0 endpoint 0x82 has invalid maxpacket 512, setting to 64
[ 3842.251625] usb 1-1: config 1 interface 0 altsetting 0 endpoint 0x5 has invalid maxpacket 512, setting to 64
[ 3842.251627] usb 1-1: config 1 interface 0 altsetting 0 endpoint 0x6 has invalid maxpacket 512, setting to 64
[ 3842.269784] usb 1-1: New USB device found, idVendor=07b8, idProduct=9271
[ 3842.269786] usb 1-1: New USB device strings: Mfr=16, Product=32, SerialNumber=48
[ 3842.269788] usb 1-1: Product: UB91C
[ 3842.269790] usb 1-1: Manufacturer: ATHEROS
[ 3842.269792] usb 1-1: SerialNumber: 12345

. my device is https://wikidevi.com/wiki/Altai_WA1011N-GU with ar9271 chipset
the driver is working perfectly on my windows xp. 
i also followed the steps to usb modeswitching at : https://wireless.wiki.kernel.org/en/users/Drivers/ath9k_htc

ifconfig is not showing the device.

**Comment by paulfertser** [comment]
On Sun, Jan 28, 2018 at 09:20:58PM +0000, DukhiAatma wrote:
> root@debian:~/Desktop# dmesg | tail
> [ 3841.784928] usb 1-1: new full-speed USB device number 6 using ohci-pci

Full-speed? Are you connecting via a non-high-speed USB hub?

> [ 3842.269784] usb 1-1: New USB device found, idVendor=07b8, idProduct=9271

You should probably try

modprobe ath9k_htc
echo 07b8 9271 | sudo tee /sys/bus/usb/drivers/ath9k_htc/new_id

-- 
Be free, use free (http://www.gnu.org/philosophy/free-sw.html) software!
mailto:fercerpav@gmail.com

**Comment by DukhiAatma** [comment]
@paulfertser  i immediately realised my mistake and changed the usb setting to 2.0 on my vm and now error messages are gone and now it is ehci-pci but the device is still not working.

**Comment by paulfertser** [comment]
On Sun, Jan 28, 2018 at 09:44:25PM +0000, DukhiAatma wrote:
> [1]@paulfertser i immediately realised my mistake and changed the usb setting to
> 2.0 on my vm and now error messages are gone and now it is ehci-pci but the
> device is still not working.

Please show dmesg after you add that vid/pid pair via new_id, as I
suggested.

-- 
Be free, use free (http://www.gnu.org/philosophy/free-sw.html) software!
mailto:fercerpav@gmail.com

**Comment by DukhiAatma** [comment]
@paulfertser  
root@debian:~# dmesg | tail
[  928.678209] ath9k_htc 1-1:1.0: FW RMW support: On
[  928.678209] ath: EEPROM regdomain: 0x32
[  928.678209] ath: EEPROM indicates we should expect a direct regpair map
[  928.678521] ath: Country alpha2 being used: 00
[  928.678526] ath: Regpair used: 0x32
[  928.696524] ieee80211 phy1: Atheros AR9271 Rev:1
[  928.728126] IPv6: ADDRCONF(NETDEV_UP): wlan0: link is not ready
[  929.510630] IPv6: ADDRCONF(NETDEV_UP): wlan0: link is not ready
[  930.574214] IPv6: ADDRCONF(NETDEV_UP): wlan0: link is not ready
[  930.617391] IPv6: ADDRCONF(NETDEV_UP): wlan0: link is not ready

**Comment by DukhiAatma** [comment]
@paulfertser  i just reconnected the device. i have to input the two commands you suggested every time but  it works  !!!!!!!

**Comment by paulfertser** [comment]
On Sun, Jan 28, 2018 at 02:09:21PM -0800, DukhiAatma wrote:
> [1]@paulfertser i just reconnected the device. i have to input the two commands
> you suggested every time but it works !!!!!!!

No, those commands should be needed just once per OS boot.

-- 
Be free, use free (http://www.gnu.org/philosophy/free-sw.html) software!
mailto:fercerpav@gmail.com

**Comment by DukhiAatma** [comment]
@paulfertser thank you for pointing the mistake. i just rebooted. and entered the commands. there was new error 

root@debian:~# dmesg | tail
[  166.509479] usb 1-1: New USB device found, idVendor=07b8, idProduct=9271
[  166.509482] usb 1-1: New USB device strings: Mfr=16, Product=32, SerialNumber=48
[  166.509484] usb 1-1: Product: UB91C
[  166.509486] usb 1-1: Manufacturer: ATHEROS
[  166.509488] usb 1-1: SerialNumber: 12345
[  210.952669] usbcore: registered new interface driver ath9k_htc
[  216.819805] usb 1-1: ath9k_htc: Firmware ath9k_htc/htc_9271-1.dev.0.fw requested
[  216.870060] usb 1-1: firmware: direct-loading firmware ath9k_htc/htc_9271-1.dev.0.fw
[  217.929258] usb 1-1: ath9k_htc: Firmware - ath9k_htc/htc_9271-1.dev.0.fw download failed
[  217.929379] usb 1-1: ath9k_htc: USB layer deinitialized

and realised i had also used compat wireless when it didn't work previously. deleted duplicate drivers and now working as you said once per boot.
